### DISCUSSION TITLE
@SerializedPrimaryKey returning Binary

### DISCUSSION THREAD

**Discussion by xenoterracide** [mikro-orm/mikro-orm#3977] [Q&A] [answered]
**Describe the bug**

`SerializedPrimaryKey` returns a `Binary` instead of a string.

**To Reproduce**

```ts
import * as uuid from 'uuid-mongodb';

@Entity({ abstract: true })
export abstract class BaseMongoAggregate<
    ID extends string,
    Event extends IEvent,
  >
{

  @PrimaryKey({ type: 'uuid' })
  protected _id: uuid.MUUID;

  constructor(id: uuid.MUUID = uuid.v4()) {
    this._id = id;
  }

  /**
   * This getter implementation is replaced by Mikro ORM
   */
  @SerializedPrimaryKey({ type: 'string' })
  get id(): ID {
    return this._id.toString() as ID;
  }
}
```

**Expected behavior**

returns a string

**Additional context**

**Versions**

```ts
❯ node --version
v18.12.1

└─ services-core@workspace:.
   └─ typescript@patch:typescript@npm%3A4.7.3#~builtin<compat/typescript>::version=4.7.3&hash=65a307 (via patch:typescript@4.7.3#~builtin<compat/typescript>)
➤ YN0013: │ @mikro-orm/mongodb@npm:5.6.8 can't be found in the cache and will be fetched from the remote registry
➤ YN0013: │ mikro-orm@npm:5.6.8 can't be found in the cache and will be fetched from the remote registry
```

**Comment by B4nan** [comment, marked as answer]
Serialized PK is for a very specific use case - when you have an objectid primary key, and it should be a regular property, not a getter. The conversion is defined in the platform. Either drop the decorator and optionally use a property serializer, or use `@Property({ persist: false })` and mark the regular pk as hidden.

**Reply by xenoterracide** [reply]
if I do that it breaks find

```
Summary of all failing tests
 FAIL  src/core/storefront-product/__tests__/storefront-product.repository.save.integration.spec.ts
  ● save StorefrontProduct › one

    NotFoundError: StorefrontProductAggregate not found ({ id: '1026f7b7-e733-4b0b-8059-d6b60f240c89' })

      106 |     expect(createdId).toBeDefined();
      107 |
    > 108 |     const found = await dao.findOneOrFail(
          |                   ^
      109 |       { id: createdId },
      110 |       {
      111 |         disableIdentityMap: true,
```

**Reply by B4nan** [reply]
Show what you did, I proposed two solutions. I am also quite sure you will need a custom mapped type if you want to use that `uuid-mongodb` package, as it looks like you want to have strings on runtime but buffers in the database.

**Reply by xenoterracide** [reply]
```ts
 @PrimaryKey({ type: 'uuid', hidden: true })
  protected _id: uuid.MUUID;

  constructor(id: uuid.MUUID = uuid.v4()) {
    super();
    this._id = id;
    const now = new Date();
    this.createdAt = now;
    this.modifiedAt = now;
  }

  /**
   * This getter implementation is replaced by Mikro ORM
   */
  @Property({ type: 'string', persist: false })
  get id(): ID {
    return this._id.toString() as ID;
  }
```

also, imo, it shouldn't really matter if it's a getter especially because it's pretty obviously an ephemeral/derived thing. Ultimately I'm just trying to do "client side" (meaning the db client) id generation.

**Reply by B4nan** [reply]
Why do you expect that to work? There is no place in your code that would convert the string id you are passing to the EM to the buffer that is stored in your database. 

This is pretty much the main reason why you can't query by getters, as there is no way to know what the getter actually does. Even if there would be a setter, we need a pure function that converts the value from your filter, not a setter that mutates some entity state. 

Normally you would create a mapped type for this and have just a single PK property - you would work with the strings as that is what the entity understands, and the ORM will convert this back and forth to the buffer variant when firing the database queries. This means you would also provide the default value as a string, not a buffer.

Similar to what [this test](https://github.com/mikro-orm/mikro-orm/blob/master/tests/issues/GH1263.test.ts) does, but for mongo you would need to have the PK mapped to `_id`.

```ts
export class UUID extends Type<string, Buffer> {

  convertToJSValue(value: Buffer) {
    return MUUID.toString(value); // just a pseudocode
  }

  convertToDatabaseValue(value: string) {
    return MUUID.fromString(value); // just a pseudocode
  }

}
```

```ts
@PrimaryKey({ name: '_id', type: UUID })
id: string = uuid();
```

**Reply by xenoterracide** [reply]
what you have with this works, but it feel like since this is really just changing to and from a mongodb binary that this could be simpler... and not require me to do this. I'm not 100% that binary is the right type, but isn't this a bit like converting `type: 'uuid'` to an actual uuid type in postgres but a varchar or something in another sql db? the `MUUID` type is simply a subtype of `Binary`

```ts
import { Type } from '@mikro-orm/core';
import { Binary } from 'mongodb';
import * as uuid from 'uuid-mongodb';

export class UuidType extends Type<string, Binary> {
  convertToJSValue(value: Binary): string {
    return value.toString();
  }

  convertToDatabaseValue(value: string): Binary {
    return uuid.from(value);
  }
}
```

**Reply by B4nan** [reply]
Postgres in particular handles this on the SQL level, you pass the string values directly in the query. AFAIK there are mysql/sqlite users who use such buffer/string types as defined in that test, sure we could provide one in the ORM codebase, but it seems trivial enough to define one that suits your needs. Note that you can as well use string uuids, both in SQL drivers and in mongo, you don't need to convert them to buffers - of course, they will have a bigger memory footprint as strings.

**Reply by xenoterracide** [reply]
and storage, not sure if the binary index's better, I'd guess yes, but I'm not sure.

**Reply by xenoterracide** [reply]
could some documentation in the mongodb section of the docs be added though please?

**Reply by ruscon** [reply]
@B4nan Thanks for `UuidType`. (Better to rename it into `MongoUuidType`)
Works perfect. 
Really needs to be added to the documentation about it.
And this:

```ts
  @PrimaryKey({ name: '_id', type: UuidType })
  id!: string
```

P.S. won't work that way
```ts
  @PrimaryKey({ name: '_id', type: () => UuidType })
  id!: string
```

**Comment by ruscon** [comment]
Found a better solution with `BSON.UUID`
If use `uuid-mongodb` in the logs then you will see `Binary.createFromBase64("oYHdWahmTUSgZPxoPf7zDQ==", 4)`
If use `BSON.UUID` in logs you will see `new UUID("9aebb91e-aa7c-4fb3-ac98-1b5b89476b9b")`

### Custom property types for `mikro-orm:

```ts
import { Type } from '@mikro-orm/core';
import { BSON } from 'mongodb';

export class MongoUuidType extends Type<string, BSON.UUID> {
  override convertToJSValue(value: BSON.UUID): string {
    return value.toString();
  }

  override convertToDatabaseValue(value: string): BSON.UUID {
    return new BSON.UUID(value);
  }
}
```

#### A string array needs a different type

```ts
import { Type } from '@mikro-orm/core';
import { BSON } from 'mongodb';

export class MongoUuidArrayType extends Type<string[] | undefined, BSON.UUID[] | undefined> {
  override convertToJSValue(values?: BSON.UUID[]): string[] | undefined {
    if (!values) {
      return values as unknown as undefined;
    }

    return values.map((value) => value.toString());
  }

  override convertToDatabaseValue(values?: string[]): BSON.UUID[] | undefined {
    if (!values) {
      return values as unknown as undefined;
    }

    return values.map((value) => new BSON.UUID(value));
  }
}
```

### Usage:

#### If the property is a primary key
```ts
@PrimaryKey({ name: '_id', type: MongoUuidType })
id!: string
```

#### If the property is a string, but not a primary key
```ts
@Property({ type: MongoUuidType })
uuid!: string
```

#### If the property is an array of strings
```ts
@Property({ type: MongoUuidArrayType })
uuids: string[]
```
